code/Model/SiteTree.php

```diff
@@ -1481,9 +1481,10 @@ public function syncLinkTracking()
         $this->extend('augmentSyncLinkTracking');
     }
 
-    public function onBeforeDelete()
+    public function onAfterDelete()
```

Comment:
As commented on #1988 

Moving the cascade delete into the "after" introduces a rather concerning shift in integrity I must point out. I can't see how this is necessary in order to simply avoid deleting all records, when it seems all that's necessary is an `isInDB()` check on children enumeration.

If you delete parents first, but fail to delete a child, you end up with your application in an inconsistent state. It needs to be leaf-first not node-first to ensure this consistency.